### ISSUE TITLE
2.2.0: lv2lint fails on plugin

### ISSUE THREAD

**Issue by dvzrv** [Yoshimi/yoshimi#164]
Hi! Similar to #125 I think this is again a problem with 2.2.0.

```
lv2lint 0.16.2
Copyright (c) 2016-2021 Hanspeter Portner (dev@open-music-kontrollers.ch)
Released under Artistic License 2.0 by Open Music Kontrollers
<http://yoshimi.sourceforge.net/lv2_plugin>
/build/.local/share/yoshimi/presets
Basic configuration /build/.config/yoshimi/yoshimi.config not found, will use default settings
/build/.local/share/yoshimi/presets
Configuration /build/.config/yoshimi/yoshimi-0.instance not found, will use default settings
Starting in LV2 plugin mode

Found 0 instruments in 0 banks
No bank 5 in root 5. Current bank is 5
Missing recent history file
    [FAIL]  Plugin Symbols
              binary exports superfluous globally visible symbols:
                * _ZNSt17_Function_handlerIFSt10unique_ptrINSt13__future_base12_Result_baseENS2_8_DeleterEEvENS1_13_State_baseV27_SetterI9PADTablesOS8_EEE10_M_managerERSt9_Any_dataRKSC_St18_Manager_operation
                * _ZTSNSt13__future_base13_State_baseV27_SetterI9PADTablesNS0_19__exception_ptr_tagEEE
                * _ZTSSt19_Sp_make_shared_tag
                * _ZZNSt9once_flag18_Prepare_executionC4IZSt9call_onceIMNSt13__future_base13_State_baseV2EFvPSt8functionIFSt10unique_ptrINS3_12_Result_baseENS7_8_DeleterEEvEEPbEJPS4_SC_SD_EEvRS_OT_DpOT0_EUlvE_EERSI_ENUlvE_4_FUNEv
                * _ZNSt17_Function_handlerIFSt10unique_ptrINSt13__future_base12_Result_baseENS2_8_DeleterEEvENS1_13_State_baseV27_SetterI9PADTablesOS8_EEE9_M_invokeERKSt9_Any_data
                * _ZNSt17_Function_handlerIFSt10unique_ptrINSt13__future_base12_Result_baseENS2_8_DeleterEEvENS1_13_State_baseV27_SetterI9PADTablesNS6_19__exception_ptr_tagEEEE9_M_invokeERKSt9_Any_data
                * _ZNSt15__exception_ptr12__dest_thunkISt12future_errorEEvPv
                * _ZTSNSt13__future_base13_State_baseV2E
                * _ZNSt13__future_base13_State_baseV29_M_do_setEPSt8functionIFSt10unique_ptrINS_12_Result_baseENS3_8_DeleterEEvEEPb
                * _ZTSSt11_Mutex_baseILN9__gnu_cxx12_Lock_policyE2EE
                * ... there is more, but the rest is being truncated
              seeAlso: <http://lv2plug.in/ns/lv2core#binary>
  <http://yoshimi.sourceforge.net/lv2_plugin#ExternalUI>
    [FAIL]  UI Symbols
              binary exports superfluous globally visible symbols:
                * _ZNSt17_Function_handlerIFSt10unique_ptrINSt13__future_base12_Result_baseENS2_8_DeleterEEvENS1_13_State_baseV27_SetterI9PADTablesOS8_EEE10_M_managerERSt9_Any_dataRKSC_St18_Manager_operation
                * _ZTSNSt13__future_base13_State_baseV27_SetterI9PADTablesNS0_19__exception_ptr_tagEEE
                * _ZTSSt19_Sp_make_shared_tag
                * _ZZNSt9once_flag18_Prepare_executionC4IZSt9call_onceIMNSt13__future_base13_State_baseV2EFvPSt8functionIFSt10unique_ptrINS3_12_Result_baseENS7_8_DeleterEEvEEPbEJPS4_SC_SD_EEvRS_OT_DpOT0_EUlvE_EERSI_ENUlvE_4_FUNEv
                * _ZNSt17_Function_handlerIFSt10unique_ptrINSt13__future_base12_Result_baseENS2_8_DeleterEEvENS1_13_State_baseV27_SetterI9PADTablesOS8_EEE9_M_invokeERKSt9_Any_data
                * _ZNSt17_Function_handlerIFSt10unique_ptrINSt13__future_base12_Result_baseENS2_8_DeleterEEvENS1_13_State_baseV27_SetterI9PADTablesNS6_19__exception_ptr_tagEEEE9_M_invokeERKSt9_Any_data
                * _ZNSt15__exception_ptr12__dest_thunkISt12future_errorEEvPv
                * _ZTSNSt13__future_base13_State_baseV2E
                * _ZNSt13__future_base13_State_baseV29_M_do_setEPSt8functionIFSt10unique_ptrINS_12_Result_baseENS3_8_DeleterEEvEEPb
                * _ZTSSt11_Mutex_baseILN9__gnu_cxx12_Lock_policyE2EE
                * ... there is more, but the rest is being truncated
              seeAlso: <http://lv2plug.in/ns/lv2core#binary>
    [WARN]  UI Instance Access
              usage of instance-access is highly discouraged
              seeAlso: <http://lv2plug.in/ns/ext/instance-access>
    [WARN]  UI Mixed DSP/UI
              mixing DSP and UI code in same binary is discouraged
              seeAlso: <http://lv2plug.in/ns/extensions/ui#>
    [WARN]  UI Toolkit
              usage of unofficial external UI is discouraged
              seeAlso: <http://kxstudio.sf.net/ns/lv2ext/external-ui#Widget>
```

**Comment by abrolag** [comment]
On Mon, 16 May 2022 09:28:06 -0700
David Runge ***@***.***> wrote:

>Hi! Similar to #125 I think this is again a problem with 2.2.0.
>
>```
>lv2lint 0.16.2
>Copyright (c) 2016-2021 Hanspeter Portner ***@***.***)
>Released under Artistic License 2.0 by Open Music Kontrollers
><http://yoshimi.sourceforge.net/lv2_plugin>
>/build/.local/share/yoshimi/presets
>Basic configuration /build/.config/yoshimi/yoshimi.config not found, will use default settings
>/build/.local/share/yoshimi/presets
>Configuration /build/.config/yoshimi/yoshimi-0.instance not found, will use default settings
>Starting in LV2 plugin mode
>
>Found 0 instruments in 0 banks
>No bank 5 in root 5. Current bank is 5
>Missing recent history file
>    [FAIL]  Plugin Symbols
>              binary exports superfluous globally visible symbols:
>                * _ZNSt17_Function_handlerIFSt10unique_ptrINSt13__future_base12_Result_baseENS2_8_DeleterEEvENS1_13_State_baseV27_SetterI9PADTablesOS8_EEE10_M_managerERSt9_Any_dataRKSC_St18_Manager_operation
>                * _ZTSNSt13__future_base13_State_baseV27_SetterI9PADTablesNS0_19__exception_ptr_tagEEE
>                * _ZTSSt19_Sp_make_shared_tag
>                * _ZZNSt9once_flag18_Prepare_executionC4IZSt9call_onceIMNSt13__future_base13_State_baseV2EFvPSt8functionIFSt10unique_ptrINS3_12_Result_baseENS7_8_DeleterEEvEEPbEJPS4_SC_SD_EEvRS_OT_DpOT0_EUlvE_EERSI_ENUlvE_4_FUNEv
>                * _ZNSt17_Function_handlerIFSt10unique_ptrINSt13__future_base12_Result_baseENS2_8_DeleterEEvENS1_13_State_baseV27_SetterI9PADTablesOS8_EEE9_M_invokeERKSt9_Any_data
>                * _ZNSt17_Function_handlerIFSt10unique_ptrINSt13__future_base12_Result_baseENS2_8_DeleterEEvENS1_13_State_baseV27_SetterI9PADTablesNS6_19__exception_ptr_tagEEEE9_M_invokeERKSt9_Any_data
>                * _ZNSt15__exception_ptr12__dest_thunkISt12future_errorEEvPv
>                * _ZTSNSt13__future_base13_State_baseV2E
>                * _ZNSt13__future_base13_State_baseV29_M_do_setEPSt8functionIFSt10unique_ptrINS_12_Result_baseENS3_8_DeleterEEvEEPb
>                * _ZTSSt11_Mutex_baseILN9__gnu_cxx12_Lock_policyE2EE
>                * ... there is more, but the rest is being truncated
>              seeAlso: <http://lv2plug.in/ns/lv2core#binary>
>  <http://yoshimi.sourceforge.net/lv2_plugin#ExternalUI>
>    [FAIL]  UI Symbols
>              binary exports superfluous globally visible symbols:
>                * _ZNSt17_Function_handlerIFSt10unique_ptrINSt13__future_base12_Result_baseENS2_8_DeleterEEvENS1_13_State_baseV27_SetterI9PADTablesOS8_EEE10_M_managerERSt9_Any_dataRKSC_St18_Manager_operation
>                * _ZTSNSt13__future_base13_State_baseV27_SetterI9PADTablesNS0_19__exception_ptr_tagEEE
>                * _ZTSSt19_Sp_make_shared_tag
>                * _ZZNSt9once_flag18_Prepare_executionC4IZSt9call_onceIMNSt13__future_base13_State_baseV2EFvPSt8functionIFSt10unique_ptrINS3_12_Result_baseENS7_8_DeleterEEvEEPbEJPS4_SC_SD_EEvRS_OT_DpOT0_EUlvE_EERSI_ENUlvE_4_FUNEv
>                * _ZNSt17_Function_handlerIFSt10unique_ptrINSt13__future_base12_Result_baseENS2_8_DeleterEEvENS1_13_State_baseV27_SetterI9PADTablesOS8_EEE9_M_invokeERKSt9_Any_data
>                * _ZNSt17_Function_handlerIFSt10unique_ptrINSt13__future_base12_Result_baseENS2_8_DeleterEEvENS1_13_State_baseV27_SetterI9PADTablesNS6_19__exception_ptr_tagEEEE9_M_invokeERKSt9_Any_data
>                * _ZNSt15__exception_ptr12__dest_thunkISt12future_errorEEvPv
>                * _ZTSNSt13__future_base13_State_baseV2E
>                * _ZNSt13__future_base13_State_baseV29_M_do_setEPSt8functionIFSt10unique_ptrINS_12_Result_baseENS3_8_DeleterEEvEEPb
>                * _ZTSSt11_Mutex_baseILN9__gnu_cxx12_Lock_policyE2EE
>                * ... there is more, but the rest is being truncated
>              seeAlso: <http://lv2plug.in/ns/lv2core#binary>
>    [WARN]  UI Instance Access
>              usage of instance-access is highly discouraged
>              seeAlso: <http://lv2plug.in/ns/ext/instance-access>
>    [WARN]  UI Mixed DSP/UI
>              mixing DSP and UI code in same binary is discouraged
>              seeAlso: <http://lv2plug.in/ns/extensions/ui#>
>    [WARN]  UI Toolkit
>              usage of unofficial external UI is discouraged
>              seeAlso: <http://kxstudio.sf.net/ns/lv2ext/external-ui#Widget>
>```

{sigh}
Thanks for letting me know.

This is most frustrating. Personally I've never been able to make sense of LV2
lint - especially, when (here at least) Yoshimi runs quite correctly as a
plugin with Ardour, Muse & Qtractor. I test them all. besides which, I can't
think of anything relevant we've changed!

I'll pass this on to the other devs, and see if someone can sort it.

-- 
Will J Godfrey
https://willgodfrey.bandcamp.com/
http://yoshimi.github.io
Say you have a poem and I have a tune.
Exchange them and we can both have a poem, a tune, and a song.

**Comment by dvzrv** [comment]
It seems that when building without link time optimization (LTO) the error does not occur.

Not entirely sure whether this can be fixed on your end. For good measure I'll ping @ventosus and @falktx here.

**Comment by falkTX** [comment]
well the usual good practices for building and linking apply here.
for plugins, make sense to only export the needed functions, so lv2lint checks for that.
this can be kinda ignored on distributions and usage where all plugins are built the same way, from the same repos and sets of libs, but as commercial vendors begin to release custom builds for linux it also becomes a problem when proper care is not taken on plugin (and host side).

so same rules as every apply:
- plugins must be self-contained, as much as possible
- do not make use of shared libs that might conflict with the host (like the usual gtk2 vs gtk3, aubio2 vs aubio3, etc)
- hide the plugin function symbols so loading a plugin does not mess with the current host process
- do not block the event loop / main thread from the plugin
- do not allocate or otherwise do non-RT stuff during audio thread

I will welcome the day where lv2lint also checks for RT safety. so that is a plugin says it is RT safe, lv2lint could check directly for that (at least the most common bad practices anyway)

coming back to the topic at hand, here are a few steps to take:
- plugin build with symbols hidden by default, unless debug build is requested
- use a linker script to ensure only specific symbols get exported
- limit the amount of shared libs used by the plugin
- if possible, separate the DSP and UI, so capable hosts can sandbox the UI and have it not conflict with the audio
- stop using fltk, as opening menus and dialogs blocks the main/current event loop

**Comment by abrolag** [comment]
On Fri, 20 May 2022 11:57:34 -0700
David Runge ***@***.***> wrote:

>It seems that when building without link time optimization (LTO) the error does not occur.
>
>Not entirely sure whether this can be fixed on your end. For good measure I'll
>ping @ventosus and @falktx here.
>

Thanks for letting me know. I'll pass it on to the others.

Unfortunately I doubt you'll get any help from falktx. In the past he's been
very hostile to Yoshimi.

-- 
Will J Godfrey
https://willgodfrey.bandcamp.com/
http://yoshimi.github.io
Say you have a poem and I have a tune.
Exchange them and we can both have a poem, a tune, and a song.

**Comment by abrolag** [comment]
Oops. Seems I was mistaken. Apologies!

-- 
Will J Godfrey
https://willgodfrey.bandcamp.com/
http://yoshimi.github.io
Say you have a poem and I have a tune.
Exchange them and we can both have a poem, a tune, and a song.

**Comment by falkTX** [comment]
I still think the amount of effort spent on Yoshimi is kinda wasted, after zynaddsubfx got a maintainer and substantial amount of work put into it.
I gave up on trying to convince anyone to merge efforts. For all I know, Mark/fundamental (current maintainer) wishes for the same, it is on Yoshimi side that things do not want to move into that direction.
Very sad state of affairs, and a bit shameful for the community in my opinion.

It is nothing personal, more about seeing projects diverge instead trying to merge instead. We have very few hands on the linux audio side, collaboration could achieve great things, but everyone prefers to do its own thing instead. Probably too late for that perhaps.

**Comment by abrolag** [comment]
Thank you for the suggestions. We will check through.

As an aside, I have no liking for FLTK, but so far we've not found an
alternative UI that readily gives the appearance and accessibility we like -
especially since we worked out scaled resizing.

-- 
Will J Godfrey
https://willgodfrey.bandcamp.com/
http://yoshimi.github.io
Say you have a poem and I have a tune.
Exchange them and we can both have a poem, a tune, and a song.

**Comment by abrolag** [comment]
I just built and ran LV2lint on the Yoshimi plugin, and see it fails claiming
a plugin from kxstudio is not valid. Indeed, on checking it seems the whole of
kxstudio has disappeared from the 'net.

So, we have a plugin that (to the best of my knowledge) works quite correctly
but is marked as faulty due to a missing external reference that plays no part
in the runtime, and we have no control over.

This is not encouraging.

-- 
Will J Godfrey
https://willgodfrey.bandcamp.com/
http://yoshimi.github.io
Say you have a poem and I have a tune.
Exchange them and we can both have a poem, a tune, and a song.

**Comment by abrolag** [comment]
On Wed, 25 May 2022 06:53:29 -0700
Filipe Coelho ***@***.***> wrote:

>Those URIs are not URLs, they do not need to point to places that actually exist.
>I have changed the domain of kxstudio, but still keep the lv2 extensions up to date to anything they might need.
>https://github.com/KXStudio/LV2-Extensions has the latest, and https://git.kx.studio/KXStudio/LV2-Extensions as mirror in case something happens to github.
>
Thanks. I've updated these, and here LV2lint is now silent.

-- 
Will J Godfrey
https://willgodfrey.bandcamp.com/
http://yoshimi.github.io
Say you have a poem and I have a tune.
Exchange them and we can both have a poem, a tune, and a song.

**Comment by kramlie** [comment]
A bit late to the party, but I'm surprised that lv2lint is complaining about globally visible symbols. We already tried to fix this [here](https://github.com/Yoshimi/yoshimi/commit/422411854164c40c26813274e4a27eb97e5ce6b6). Is arch ignoring our build arguments and using their own? It needs to pass the `-fvisibility=hidden` argument.

IIRC, the symbols it's talking about are the "GLOBAL DEFAULT" ones. On my system I get only these:

```
$ readelf -s LV2_Plugin/yoshimi_lv2.so | grep 'GLOBAL *DEFAULT' | grep -v ' UND '
   491: 00000000005d7420    23 FUNC    GLOBAL DEFAULT   14 lv2ui_descriptor
   509: 00000000005d73f0    35 FUNC    GLOBAL DEFAULT   14 lv2_descriptor
   528: 00000000009a2350     0 NOTYPE  GLOBAL DEFAULT   26 _edata
   543: 00000000009a2350     0 NOTYPE  GLOBAL DEFAULT   27 __bss_start
   549: 0000000000dda968     0 NOTYPE  GLOBAL DEFAULT   27 _end
 13632: 00000000009a2350     0 NOTYPE  GLOBAL DEFAULT   26 _edata
 13665: 00000000005d73f0    35 FUNC    GLOBAL DEFAULT   14 lv2_descriptor
 13770: 00000000005d7420    23 FUNC    GLOBAL DEFAULT   14 lv2ui_descriptor
 13973: 0000000000dda968     0 NOTYPE  GLOBAL DEFAULT   27 _end
 14018: 00000000009a2350     0 NOTYPE  GLOBAL DEFAULT   27 __bss_start
```

In other words not much, and only two function definitions, the two that we want.
